File: backend/services/vote-service/logger/logger.go
package logger

import (
	"context"
	"fmt"
	"log"
)

// Info logs a message with the correlation ID from the context if it exists.
func Info(ctx context.Context, format string, v ...interface{}) {
	msg := fmt.Sprintf(format, v...)
	logWithContext(ctx, "INFO", msg)
}

// Error logs an error message with the correlation ID from the context.
func Error(ctx context.Context, format string, v ...interface{}) {
	msg := fmt.Sprintf(format, v...)
	logWithContext(ctx, "ERROR", msg)
}

// Warn logs a warning message with the correlation ID from the context.
func Warn(ctx context.Context, format string, v ...interface{}) {
	msg := fmt.Sprintf(format, v...)
	logWithContext(ctx, "WARN", msg)
}

func logWithContext(ctx context.Context, level, msg string) {
	var traceId string
	if ctx != nil {
		if val := ctx.Value("correlation_id"); val != nil {
			traceId = val.(string)
		}
	}
	if traceId == "" {
		traceId = "NO-TRACE"
	}
	log.Printf("[Trace: %s] [%s] %s\n", traceId, level, msg)
}

File: backend/services/realtime-service/src/utils/logger.js
import { AsyncLocalStorage } from 'async_hooks';

export const traceStorage = new AsyncLocalStorage();

const logWithTrace = (level, ...args) => {
  const traceId = traceStorage.getStore();
  const tracePrefix = traceId ? `[Trace: ${traceId}]` : '[Trace: NO-TRACE]';
  const timestamp = new Date().toISOString();
  
  if (level === 'error') {
    console.error(`[${timestamp}] [ERROR] ${tracePrefix}`, ...args);
  } else if (level === 'warn') {
    console.warn(`[${timestamp}] [WARN] ${tracePrefix}`, ...args);
  } else {
    console.log(`[${timestamp}] [INFO] ${tracePrefix}`, ...args);
  }
};

export const logger = {
  info: (...args) => logWithTrace('info', ...args),
  error: (...args) => logWithTrace('error', ...args),
  warn: (...args) => logWithTrace('warn', ...args),
};
